commit a8d21a465dfe99e3627f0649a2f97abcd4776c92
Date:   Mon Feb 8 13:48:10 2010 +0100

    Bug #593176 - winmail.dat sent in outgoing plaintext mail

diff -upr evolutionmapi.orig/src/camel/camel-mapi-utils.c evolutionmapi/src/camel/camel-mapi-utils.c
--- evolutionmapi.orig/src/camel/camel-mapi-utils.c	2010-06-23 12:37:47.000000000 +0200
+++ evolutionmapi/src/camel/camel-mapi-utils.c	2010-06-23 12:38:03.000000000 +0200
@@ -68,6 +68,7 @@ mapi_item_add_recipient (const char *rec
 {
 	ExchangeMAPIRecipient *recipient;
 	uint32_t val = 0;
+	uint8_t bVal;
 	const char *str = NULL;
 
 	if (!recipients)
@@ -88,7 +89,7 @@ mapi_item_add_recipient (const char *rec
 	set_SPropValue_proptag (&(recipient->in.req_lpProps[1]), PR_SEND_INTERNET_ENCODING, (const void *)&val);
 
 	/* External recipient properties - set them only when the recipient is unresolved */
-	recipient->in.ext_cValues = 7;
+	recipient->in.ext_cValues = 8;
 	recipient->in.ext_lpProps = g_new0 (struct SPropValue, recipient->in.ext_cValues + 1);
 
 	val = DT_MAILUSER;
@@ -104,6 +105,9 @@ mapi_item_add_recipient (const char *rec
 	set_SPropValue_proptag (&(recipient->in.ext_lpProps[5]), PR_DISPLAY_NAME, (const void *)(str));
 	set_SPropValue_proptag (&(recipient->in.ext_lpProps[6]), PR_7BIT_DISPLAY_NAME, (const void *)(str));
 
+	bVal = FALSE;
+	set_SPropValue_proptag (&(recipient->in.ext_lpProps[7]), PR_SEND_RICH_INFO, &bVal);
+
 	*recipient_list = g_slist_append (*recipient_list, recipient);
 }
 
