diff -up evolution-mapi-0.28.3/src/camel/camel-mapi-utils.c.extra-re-in-replies evolution-mapi-0.28.3/src/camel/camel-mapi-utils.c
--- evolution-mapi-0.28.3/src/camel/camel-mapi-utils.c.extra-re-in-replies	2010-03-01 08:28:12.000000000 -0500
+++ evolution-mapi-0.28.3/src/camel/camel-mapi-utils.c	2010-06-12 17:17:52.833528162 -0400
@@ -409,8 +409,9 @@ camel_mapi_utils_create_item_build_props
 	*cpid = 65001; /* UTF8 */
 	set_SPropValue_proptag(&props[i++], PR_INTERNET_CPID, cpid);
 	set_SPropValue_proptag(&props[i++], PR_SUBJECT_UNICODE, g_strdup (item->header.subject));
-	set_SPropValue_proptag(&props[i++], PR_CONVERSATION_TOPIC_UNICODE, g_strdup (item->header.subject));
-	set_SPropValue_proptag(&props[i++], PR_NORMALIZED_SUBJECT_UNICODE, g_strdup (item->header.subject));
+	/* PR_CONVERSATION_TOPIC_UNICODE and PR_NORMALIZED_SUBJECT_UNICODE,
+	 * together with PR_SUBJECT_PREFIX_UNICODE are computed from
+	 * PR_SUBJECT by the server. */
 
 	*send_rich_info = false;
 	set_SPropValue_proptag(&props[i++], PR_SEND_RICH_INFO, (const void *) send_rich_info);
